Automattic/jetpack
# JITMs: add component to be used in the Jetpack Dashboard.

#### Changes proposed in this Pull Request:

JITMs: add component to be used in the Jetpack Dashboard.

<img width="868" alt="screenshot 2018-12-05 at 17 17 41" src="https://user-images.githubusercontent.com/426388/49527273-db7b7080-f8b1-11e8-8b53-f3cd6a41cc2a.png">
<img width="835" alt="screenshot 2018-12-05 at 17 17 29" src="https://user-images.githubusercontent.com/426388/49527275-dc140700-f8b1-11e8-8ae9-c16768431556.png">
<img width="828" alt="screenshot 2018-12-05 at 17 19 12" src="https://user-images.githubusercontent.com/426388/49527297-e7673280-f8b1-11e8-98af-669bd12946d6.png">
<img width="897" alt="screenshot 2018-12-06 at 11 55 06" src="https://user-images.githubusercontent.com/426388/49579890-d3710e80-f94d-11e8-9a14-160a06a06bd2.png">
<img src="https://user-images.githubusercontent.com/426388/49581453-509e8280-f952-11e8-9561-783cad310e91.gif">

The first part of this PR (the necessary new method in the rest api client and the new redux structure) was merged in #10818. This PR brings the second part: a new Jitm component and a new QueryJitm component, as well as a call to the new Jitm component in the main dashboard.

- [x] Check all the to-dos for things to do. 
- [x] Dequeue regular JITMs on the JP dashboard pages.
- [x] Figure out what to do with existing JITMs; with the current implementation we have no JITMs that are set to be displayed on JP dash. We will most likely need to edit the JITMs on wpcom to add the string matching JP Dash for some messages.
- [x] Some HTML encoding issues with the messages.

#### Testing instructions:

* You'll need a WordPress.com sandbox.
* Start by applying D19944-code.
* Sandbox your Jetpack site with the `JETPACK__SANDBOX_DOMAIN` constant ([reference](https://github.com/Automattic/jetpack/blob/add/jitm-component/docker/README.md#must-use-plugins-directory)). 
    - If you don't have a WordPress.com sandbox, you can sandbox mine. Through a Jurassic Ninja site with that branch, go to Settings > Jetpack Constants, and add `sandboxme.wordpress.com` as the sandbox you want to use.
* Visit Jetpack > Dashboard: a message should appear.
* Visit Jetpack > Settings: the message should disappear.
* Visit Jetpack > Plans: a different message should appear.

#### Known issues / To do:

- `toplevel_page_jetpack` should continue to work as a string one can use to target **all** pages of the Jetpack dashboard at once.
- ~Lists are not displayed properly right now.~
- ~Props warnings for `isActivatingModule` and `isModuleActivated`~
- ~When dismissing a message, and then switching to a different tab, the dismissed message is visible for a few seconds until the new message (or no message at all) is loaded.~
- ~When activating a module via a JITM, the JITM remains in place once the module has been activated. This is related to #9462~ I opted to remove the button once the module has been activated. This way one can still click on the CTA to learn more if they are interested. If they leave the page or refresh, the JITM will disappear for good.
- ~The look of the messages is a bit different from the ones you'll find in other places in wp-admin. The CTA button uses Calypso's button component, and as such the text inside the button is in uppercase. That is not the case with the wp-admin implementation of the button.~

#### Proposed changelog entry for your changes:

* Notices: display more relevant notices depending on the Jetpack dashboard screen you visit.
